core/src/service/search_service.rs

```diff
@@ -49,4 +67,346 @@ impl RequestContext<'_, '_> {
+                            .map(Some)?,
+                        _ => None,
+                    };
+
+                files_info.push(SearchableFileInfo {
+                    id: file.id,
+                    path: RequestContext::path_by_id_helper(&files_not_deleted, file.id)?,
+                    content,
+                })
+            }
+        }
+
+        Ok(thread::spawn(move || {
+            if let Err(search_err) = Self::search(&results_tx, search_rx, files_info) {
+                if let Err(err) = results_tx.send(SearchResult::Error(search_err)) {
+                    warn!("Send failed: {:#?}", err);
+                }
+            }
+        }))
+    }
+
+    fn search(
+        results_tx: &Sender<SearchResult>, search_rx: Receiver<SearchRequest>,
+        files_info: Vec<SearchableFileInfo>,
+    ) -> Result<(), UnexpectedError> {
+        let files_info = Arc::new(files_info);
+        let mut should_continue = Arc::new(AtomicBool::new(true));
+        let mut last_search = None;
+
+        loop {
+            let search = match last_search {
+                None => match search_rx.try_iter().last() {
+                    Some(search) => search,
+                    None => match search_rx.recv() {
+                        Ok(search) => search,
+                        Err(_) => return Ok(()),
+                    },
+                },
+                Some(search) => {
+                    last_search = None;
+                    search
+                }
+            };
+
+            thread::sleep(Duration::from_millis(DEBOUNCE_MILLIS));
+
+            match search_rx.try_iter().last() {
+                None => {
+                    match &search {
+                        SearchRequest::Search { input } => {
+                            should_continue.store(false, atomic::Ordering::Relaxed);
+                            should_continue = Arc::new(AtomicBool::new(true));
+
+                            let results_tx = results_tx.clone();
+                            let files_info = files_info.clone();
+                            let should_continue = should_continue.clone();
+                            let input = input.clone();
+
+                            thread::spawn(move || {
+                                if let Err(search_err) = RequestContext::search_loop(
+                                    &results_tx,
+                                    files_info,
+                                    should_continue,
+                                    input,
+                                ) {
+                                    if let Err(err) =
+                                        results_tx.send(SearchResult::Error(search_err))
+                                    {
+                                        warn!("Send failed: {:#?}", err);
+                                    }
+                                }
+                            });
```

Comment:
What happens if the last search is still going on?

(I don't think this is an actual problem because of debounce, and how fast you say search results come in)

But I think it points to a larger data modeling issue.

Reply:
I think you need to give some thought to:
1. how we can detect if a new search is stepping on an old search
2. what should the model of cancelling searches be? I get the sense there should be some vec of atomic bools or something like that. You want to be able to cancel all prior searches, there could be many.

You can test for all these things by adding sleeps.